[BUG] Limited API bytes.decode fails

### Describe the bug

Thanks for all the hard work on the Py_LIMITED_API :heart: 

I tried compiling one of my projects with `3.0.1` and ran into a few issues:

The following produces an error when trying to compile with the Limited API enabled and cython `language="c++"`:

```
b"123".decode("utf-8")
```

Here is a full example: https://github.com/ap--/seabreeze-c-backend/tree/cython-limitedapi-errors/cython_pylimited_errors

Error message:

```
building 'limitedtest.bytes_decode' extension
x86_64-linux-gnu-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DCYTHON_LIMITED_API=1 -DPy_LIMITED_API=0x030B0000 -I/tmp/build-env-f2q3qnhc/include -I/usr/include/python3.11 -c limitedtest/bytes_decode.cpp -o build/temp.linux-x86_64-cpython-311/limitedtest/bytes_decode.o
limitedtest/bytes_decode.cpp: In function ‘PyObject* __Pyx_PyCode_New(int, int, int, int, int, int, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, int, PyObject*)’:
limitedtest/bytes_decode.cpp:610:19: warning: unused variable ‘version_info’ [-Wunused-variable]
  610 |         PyObject *version_info; // borrowed
      |                   ^~~~~~~~~~~~
limitedtest/bytes_decode.cpp: In function ‘PyObject* __Pyx_decode_bytes(PyObject*, Py_ssize_t, Py_ssize_t, const char*, const char*, PyObject* (*)(const char*, Py_ssize_t, const char*))’:
limitedtest/bytes_decode.cpp:1562:9: error: ‘PyBytes_AS_STRING’ was not declared in this scope; did you mean ‘PyBytes_AsString’?
 1562 |         PyBytes_AS_STRING(string), PyBytes_GET_SIZE(string),
      |         ^~~~~~~~~~~~~~~~~
      |         PyBytes_AsString
limitedtest/bytes_decode.cpp:1562:36: error: ‘PyBytes_GET_SIZE’ was not declared in this scope; did you mean ‘PyBytes_Size’?
 1562 |         PyBytes_AS_STRING(string), PyBytes_GET_SIZE(string),
      |                                    ^~~~~~~~~~~~~~~~
      |                                    PyBytes_Size
limitedtest/bytes_decode.cpp: In function ‘PyObject* __Pyx_PyUnicode_FromString(const char*)’:
limitedtest/bytes_decode.cpp:3728:65: error: ‘strlen’ was not declared in this scope
 3728 |     return __Pyx_PyUnicode_FromStringAndSize(c_str, (Py_ssize_t)strlen(c_str));
      |                                                                 ^~~~~~
limitedtest/bytes_decode.cpp:1398:80: note: in definition of macro ‘__Pyx_PyUnicode_FromStringAndSize’
 1398 | #define __Pyx_PyUnicode_FromStringAndSize(c_str, size) PyUnicode_Decode(c_str, size, __PYX_DEFAULT_STRING_ENCODING, NULL)
      |                                                                                ^~~~
limitedtest/bytes_decode.cpp:2698:1: note: ‘strlen’ is defined in header ‘<cstring>’; did you forget to ‘#include <cstring>’?
 2697 | #include "traceback.h"
  +++ |+#include <cstring>
 2698 | #if PY_VERSION_HEX >= 0x030b00a6 && !CYTHON_COMPILING_IN_LIMITED_API
error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1


```


### Code to reproduce the behaviour:

```cython
x = b"123".decode("utf-8")
```


### Expected behaviour

_No response_

### OS

Linux

### Python version

3.11

### Cython version

3.0.1

### Additional context

_No response_

--- a/Cython/Utility/StringTools.c
+++ b/Cython/Utility/StringTools.c
@@ -596,8 +596,18 @@ static CYTHON_INLINE PyObject* __Pyx_decode_bytes(
          PyObject* string, Py_ssize_t start, Py_ssize_t stop,
          const char* encoding, const char* errors,
          PyObject* (*decode_func)(const char *s, Py_ssize_t size, const char *errors)) {
+    char* as_c_string;
+    Py_ssize_t size;
+#if CYTHON_ASSUME_SAFE_MACROS
+    as_c_string = PyBytes_AS_STRING(string);
+    size = PyBytes_GET_SIZE(string);
+#else
+    if (PyBytes_AsStringAndSize(string, &as_c_string, &size) < 0) {
+        return NULL;
+    }
+#endif
     return __Pyx_decode_c_bytes(
-        PyBytes_AS_STRING(string), PyBytes_GET_SIZE(string),
+        as_c_string, size,
         start, stop, encoding, errors, decode_func);
 }
 
@@ -608,8 +618,17 @@ static CYTHON_INLINE PyObject* __Pyx_decode_bytearray(
          PyObject* string, Py_ssize_t start, Py_ssize_t stop,
          const char* encoding, const char* errors,
          PyObject* (*decode_func)(const char *s, Py_ssize_t size, const char *errors)) {
+    char* as_c_string;
+    Py_ssize_t size;
+#if CYTHON_ASSUME_SAFE_MACROS
+    as_c_string = PyBytes_AS_STRING(string);
+    size = PyBytes_GET_SIZE(string);
+#else
+    if (!(as_c_string = PyByteArray_AsString(string))) return NULL;
+    if ((size = PyByteArray_Size(string)) < 0) return NULL;
+#endif
     return __Pyx_decode_c_bytes(
-        PyByteArray_AS_STRING(string), PyByteArray_GET_SIZE(string),
+        as_c_string, size,
         start, stop, encoding, errors, decode_func);
 }
 

--- a/Cython/Utility/StringTools.c
+++ b/Cython/Utility/StringTools.c
@@ -621,8 +621,8 @@ static CYTHON_INLINE PyObject* __Pyx_decode_bytearray(
     char* as_c_string;
     Py_ssize_t size;
 #if CYTHON_ASSUME_SAFE_MACROS
-    as_c_string = PyBytes_AS_STRING(string);
-    size = PyBytes_GET_SIZE(string);
+    as_c_string = PyByteArray_AS_STRING(string);
+    size = PyByteArray_GET_SIZE(string);
 #else
     if (!(as_c_string = PyByteArray_AsString(string))) return NULL;
     if ((size = PyByteArray_Size(string)) < 0) return NULL;

--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -1,6 +1,7 @@
 /////////////// TypeConversions.proto ///////////////
 
 /* Type Conversion Predeclarations */
+#include <string.h>
 
 #define __Pyx_uchar_cast(c) ((unsigned char)c)
 #define __Pyx_long_cast(x) ((long)x)

--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -1,7 +1,7 @@
 /////////////// TypeConversions.proto ///////////////
+//@requires: StringTools.c::IncludeStringH
 
 /* Type Conversion Predeclarations */
-#include <string.h>
 
 #define __Pyx_uchar_cast(c) ((unsigned char)c)
 #define __Pyx_long_cast(x) ((long)x)

--- a/tests/run/isolated_limited_api_tests.srctree
+++ b/tests/run/isolated_limited_api_tests.srctree
@@ -33,6 +33,8 @@ limited.C()
 limited.D()
 limited.E()
 
+assert limited.decode(b'a', bytearray(b'b')) == "ab"
+
 
 ##################### limited.pyx #############################
 
@@ -52,6 +54,9 @@ def fib(int n):
 def raises():
     raise RuntimeError()
 
+def decode(bytes b, bytearray ba):
+    return b.decode("utf-8") + ba.decode("utf-8")
+    
 class C:
     pass
 

